elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/Schema.java

```diff
@@ -0,0 +1,250 @@
+import com.yahoo.elide.datastores.aggregation.annotation.Meta;
+import com.yahoo.elide.datastores.aggregation.annotation.MetricAggregation;
+import com.yahoo.elide.datastores.aggregation.annotation.MetricComputation;
+import com.yahoo.elide.datastores.aggregation.annotation.Temporal;
+import com.yahoo.elide.datastores.aggregation.dimension.DegenerateDimension;
+import com.yahoo.elide.datastores.aggregation.dimension.Dimension;
+import com.yahoo.elide.datastores.aggregation.dimension.EntityDimension;
+import com.yahoo.elide.datastores.aggregation.dimension.TimeDimension;
+import com.yahoo.elide.datastores.aggregation.metric.AggregatedMetric;
+import com.yahoo.elide.datastores.aggregation.metric.Aggregation;
+import com.yahoo.elide.datastores.aggregation.metric.Metric;
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TimeZone;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * {@link Schema} keeps track of table, metrics, and dimensions of an entity for AggregationDataStore.
+ * <p>
+ * On calling {@link DataStore#populateEntityDictionary(EntityDictionary)}, one {@link Schema} will be created for each
+ * entity.
+ * <p>
+ * By overriding {@link #constructDimension(String, Class, EntityDictionary)} and
+ * {@link #constructMetric(String, Class, EntityDictionary)}, people can have new schema backed by their own defined
+ * {@link Dimension}s and {@link Metric}s.
+ * <p>
+ * {@link Schema} is thread-safe and can be accessed by multiple-threads.
+ */
+@Slf4j
+public class Schema {
+
+    @Getter
+    private final Class<?> entityClass;
+    @Getter
+    private final Set<Metric> metrics;
+    @Getter
+    private final Set<Dimension> dimensions;
+    @Getter
+    private final Set<Column> columns;
+    @Getter(value = AccessLevel.PRIVATE)
+    private final EntityDictionary entityDictionary;
+
+    /**
+     * Constructor
+     * <p>
+     * This constructor calls {@link #constructDimension(String, Class, EntityDictionary)} and
+     * {@link #constructMetric(String, Class, EntityDictionary)} ()} to construct all {@link Dimension}s and
+     * {@link Metric}s associated with the entity class passed in.
+     *
+     * @param cls  The type of the entity, whose {@link Schema} is to be constructed
+     * @param entityDictionary  The meta info object that helps to construct {@link Schema}
+     *
+     * @throws NullPointerException if anyone of the arguments is {@code null}
+     */
+    public Schema(Class<?> cls, EntityDictionary entityDictionary) {
+        this.entityClass = Objects.requireNonNull(cls, "cls");
+        this.entityDictionary = Objects.requireNonNull(entityDictionary, "entityDictionary");
+
+        this.metrics = entityDictionary.getAllFields(getEntityClass()).stream()
```

Comment:
I preferred the old way where we had methods getAllMetrics and getAllDimensions.  BetterCodeHub will probably ding this for complexity and number of lines.